cur-test/cur/tests/stlc-formal.rkt

```diff
@@ -0,0 +1,344 @@
+
+; notB = \x:Bool. test x then fls else tru
+(define notB (abs x Bool (test (var x) fls tru)))
+
+;; =======================
+;;  OPERATIONAL SEMANTICS
+;; =======================
+
+;; --------
+;;  Values
+;; --------
+(define-datatype value : (-> tm Prop)
+  [v-abs : (∀ [x : String] [T : ty] [t : tm] (value (abs x T t)))]
+  [v-tru : (value tru)]
+  [v-fls : (value fls)])
+
+;; --------------
+;;  Substitution
+;; --------------
+
+;; TODO
+;; Technical note: Substitution becomes trickier to define if we consider the
+;; case where s, the term being substituted for a variable in some other term,
+;; may itself contain free variables.
+(define/rec/match subst-tmp : [x : String] [s : tm] tm -> tm
+  [(var xp) => (if (string-equal? x xp)
+                   s
+                   (var xp))]
+  [(abs xp T t1) => (if (string-equal? x xp)
+                        (abs xp T t1)
+                        (abs xp T (subst-tmp x s t1)))]
+  [tru => tru]
+  [fls => fls]
+  [(test t1 t2 t3) => (test (subst-tmp x s t1) (subst-tmp x s t2) (subst-tmp x s t3))])
+
+(define subst subst-tmp)
+
+;; -----------
+;;  Reduction
+;; -----------
+(define-datatype step : (-> tm tm Prop)
+  [ST-AppAbs : (∀ [x : String] [T : ty] [t12 : tm] [v2 : tm]
+                  (-> (value v2)
+                      (step (app (abs x T t12) v2) (subst x v2 t12))))]
+  [ST-App1 : (∀ [t1 : tm] [t1p : tm] [t2 : tm]
+                (-> (step t1 t1p)
+                    (step (app t1 t2) (app t1p t2))))]
+  [ST-App2 : (∀ [v1 : tm] [t2 : tm] [t2p : tm]
+                (-> (value v1)
+                    (step t2 t2p)
+                    (step (app v1 t2) (app v1 t2p))))]
+  [ST-TestTru : (∀ [t1 : tm] [t2 : tm]
+                   (step (test tru t1 t2) t1))]
+  [ST-TestFls : (∀ [t1 : tm] [t2 : tm]
+                   (step (test fls t1 t2) t2))]
+  [ST-Test : (∀ [t1 : tm] [t1p : tm] [t2 : tm] [t3 : tm]
+                (-> (step t1 t1p)
+                    (step (test t1 t2 t3) (test t1p t2 t3))))])
+
+(define-syntax -->
+  (syntax-parser
+    [(_ t1 t2) #'(step t1 t2)]))
+
+(define-syntax -->*
+  (syntax-parser
+    [(_ t1 t2) #'((multi tm step) t1 t2)]))
+
+;; ----------
+;;  Examples
+;; ----------
+
+; (\x:Bool→Bool. x) (\x:Bool. x) -->* \x:Bool. x
+#;(define-theorem step-example-1
+  (-->* (app idBB idB) idB)
+    ; TODO
+    ; https://coq.inria.fr/refman/proof-engine/tactics.html#coq:tacv.eapply
+    ; The tactic eapply behaves like apply but it does not fail when no
+    ; instantiations are deducible for some variables in the premises. Rather,
+    ; it turns these variables into existential variables which are variables
+    ; still to instantiate (see Existential variables). The instantiation is
+    ; intended to be found later in the proof.
+  (by-apply multi-step) ; by-apply: could not infer instantiation of: multi-step; try explicit #:with or #:with-var args?
+  (by-apply ST-AppAbs)
+  (by-apply v-abs)
+  simpl
+  (by-apply multi-refl))
+

```

Comment:
So none of the STLC examples are working? Did you try to give `by-apply` the explicit arguments from the explicit proof term below?

Reply:
I asked Paul to put the STLC test in there, even though most of the tactic proofs don't work yet. We could make it work manually, but this formalism was a direct port from a Coq tutorial. Even so, the large formalism and things that do work are good tests, and are slow, so a good benchmark.

Reply:
Ok cool. I'll try to check it out later to see if there's an obvious fix.

Reply:
The problem is that we don't implement `eapply` and the Coq proofs used that to automagically figure out the arguments. We could probably fix is pretty quickly as you suggested.